src/Passes/Source/Commandline/ParameterParser.hpp

```diff
@@ -0,0 +1,105 @@
+#pragma once
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include <string>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+namespace microsoft
+{
+namespace quantum
+{
+
+    /// Parameter parser class which allows the developer to specify a set of default settings and
+    /// update those using the commandline argc and argv.
+    class ParameterParser
+    {
+      public:
+        using String      = std::string;
+        using Arguments   = std::vector<String>;
+        using Flags       = std::unordered_set<String>;
+        using SettingsMap = std::unordered_map<String, String>;
+
+        // Construction and deconstrution configuration
+        //
+
+        /// Parameter parsers requires a setting class to store
+        /// parameters passed. The parameter parser takes a set of
+        /// default Settings as its first argument.
+        ParameterParser() = default;
+
+        // No copy construction.
+        ParameterParser(ParameterParser const& other) = delete;
+
+        // Allow move semantics.
+        ParameterParser(ParameterParser&& other) = default;
+
+        // Default destruction.
+        ~ParameterParser() = default;
+
+        // Configuration
+        //
+
+        /// Marks a name as a flag (as opposed to an option).
+        /// This ensures that no parameter is expected after
+        /// the flag is specified. For instance `--debug` is
+        /// a flag as opposed to `--log-level 3` which is an
+        /// option.
+        void addFlag(String const& v);
+
+        // Operation
+        //
+
+        /// Parses the command line arguments given the argc and argv
+        /// from the main function.
+        void parseArgs(int argc, char** argv);
+
+        /// Returns list of arguments without flags and/or options
+        /// included.
+        Arguments const& arguments() const;
+
+        /// Returns the n'th commandline argument.
+        String const& getArg(uint64_t const& n);
+
+        /// Gets a named setting, falling back to a default if the key is not found.
+        String get(String const& name, String const& default_value) const noexcept;
+
+        /// Gets a named setting. This method throws if the setting is not present.
+        String get(String const& name) const;
```

Comment:
Did you consider returning a reference to const String (`String const&`) rather than a copy? Returning a copy results in a dynamic allocation and deep copy in common case.